trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,332 @@
+
+/// A backing-storage independent immutable representation of a GraphQL type.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Type {
+    base: Arc<str>,
+    modifiers: Modifiers,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Modifiers {
+    mask: u64, // space for ~30 levels of list nesting
+}
+
+impl Modifiers {
+    const NON_NULLABLE_MASK: u64 = 1;
+    const LIST_MASK: u64 = 2;
+    const MAX_LIST_DEPTH: u64 = 30;
+    // represents the most far left list bit that can be set before adding a new list will overflow
+    const MAX_LIST_DEPTH_MASK: u64 = Self::LIST_MASK << ((Self::MAX_LIST_DEPTH - 1) * 2); // - 1 because we start shifted over once.
+
+    fn nullable(&self) -> bool {
+        (self.mask & Self::NON_NULLABLE_MASK) == 0
+    }
+
+    fn is_list(&self) -> bool {
+        (self.mask & Self::LIST_MASK) != 0
+    }
+
+    fn as_list(&self) -> Option<Modifiers> {
+        self.is_list().then_some(Modifiers { mask: self.mask >> 2 })
+    }
+
+    fn at_max_list_depth(&self) -> bool {
+        (self.mask & (Self::MAX_LIST_DEPTH_MASK)) == (Self::MAX_LIST_DEPTH_MASK)
+    }
+}
+
+impl Type {
+    /// Creates a new [`Type`] from a string.
+    /// Returns `None` if the string is not a valid GraphQL type.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let ty = Type::new("[String!]!").unwrap();
+    /// assert_eq!(ty.to_string(), "[String!]!");
+    ///
+    /// assert_eq!(Type::new("[String!]").unwrap().to_string(), "[String!]");
+    /// ```
+    pub fn new(ty: &str) -> Option<Self> {
+        Some(Self::from_type(&GQLType::new(ty)?))
+    }
+
+    /// Creates an individual [`Type`], not a list.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let nullable = false;
+    /// let ty = Type::new_named_type("String", nullable);
+    ///
+    /// assert_eq!(ty.to_string(), "String!");
+    /// assert_eq!(ty, Type::new("String!").unwrap());
+    /// ```
+    pub fn new_named_type(base_type_name: &str, nullable: bool) -> Self {
+        Self {
+            base: base_type_name.to_string().into(),
+            modifiers: Modifiers { mask: if nullable { 0 } else { Modifiers::NON_NULLABLE_MASK } },
+        }
+    }
+
+    /// Creates a new list layer on a [`Type`].
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let inner_nullable = false;
+    /// let inner_ty = Type::new_named_type("String", inner_nullable);
+    ///
+    /// let outer_nullable = true;
+    /// let ty = Type::new_list_type(inner_ty, outer_nullable);
+    ///
+    /// assert_eq!(ty.to_string(), "[String!]");
+    /// assert_eq!(ty, Type::new("[String!]").unwrap());
+    /// ```
+    pub fn new_list_type(inner_type: Self, nullable: bool) -> Self {
+        if inner_type.modifiers.at_max_list_depth() {
+            panic!("too many nested lists");
```

Comment:
Panics with a constant error message are generally best avoided because they make debugging hard.

Try to include relevant information about the exact inputs that caused this.